### PR TITLE
Events

### PR DESCRIPTION
**Pull request by LoganWright**
Initial stab at basic events hub, let's talk about it and I'll add readme

### CODE REVIEW

File: Sources/Vapor/Events/Events.swift [hunk 1]
```diff
@@ -0,0 +1,92 @@
+/**
+ This class is used to keep a subscription to an event active.
+ 
+ Set all references to `nil` to no longer receive events
+ */
+public final class Subscription {
+    
+    /**
+     Completiont to run on deinit.
```
Comment (tannernelson) [inline Sources/Vapor/Events/Events.swift]: sp

File: Vapor.xcodeproj/project.xcworkspace/xcshareddata/Vapor.xcscmblueprint [hunk 2]
```diff
@@ -50,17 +46,12 @@
...
-      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryIdentifierKey" : "48D51A061B4CAA12BB1778714EEA3375428E1708"
-    },
-    {
       "DVTSourceControlWorkspaceBlueprintRemoteRepositoryURLKey" : "https:\/\/github.com\/qutheory\/SHA1.git",
       "DVTSourceControlWorkspaceBlueprintRemoteRepositorySystemKey" : "com.apple.dt.Xcode.sourcecontrol.Git",
       "DVTSourceControlWorkspaceBlueprintRemoteRepositoryIdentifierKey" : "4A88C7F5F6EA13C3C2A03CD1950D7F655F1E4F26"
     },
     {
-      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryURLKey" : "github.com:qutheory\/vapor.git",
+      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryURLKey" : "github.com:LoganWright\/vapor.git",
```
Comment (tannernelson) [inline Vapor.xcodeproj/project.xcworkspace/xcshareddata/Vapor.xcscmblueprint]: what's that

**Comment by tannernelson** [comment]
This is awesome. Can you try committing again to see if Hound CI can get it's life together?

File: Vapor.xcodeproj/project.xcworkspace/xcshareddata/Vapor.xcscmblueprint (hunk 2 above)
Comment (LoganWright) [inline Vapor.xcodeproj/project.xcworkspace/xcshareddata/Vapor.xcscmblueprint]: I don't know, I saw that too! I have a repo forked on my personal account that I use occasionally for pre tagging, etc. It must've picked that up! I'll swap it back.

**Comment by shnhrrsn** [comment]
I don't love having to retain subscriptions, I'd like to see if we could find a way around it.  Maybe something like:

```swift
BatteryEvent.subscribe { [weak self] level in
   guard let s = self else {
      return false
   }

   // Handle event

   return true
 }
```

Returning false would trigger an unsubscribe when `BatteryEvent` is published.  Not sold on this either, just wanted to throw an alternative out there and start a dialog.

**Comment by tannernelson** [comment]
I kind of like the approach. As long as you hold that subscription, you are subscribed to the event. To unsubscribe, just deallocate it. It's definitely Swifty, and an elegant way to ensure that as any of your data structures holding subscriptions get deallocated, their subscriptions end as well.

**Comment by LoganWright** [comment]
@shnhrrsn I think there's very likely a way to build a syntax similar to what you want on top of what's existing here. Let me see if I can whip up an example.

File: Sources/Vapor/Events/Events.swift [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    
+    /**
+     Adds a subscriber for this event with a handler to fire on post.
+     
+     - Warning: subscription returned from this function must be retained to receive events
+     
+     - returns: a subscription. As long as it's retained, the passed handler will fire
+     */
+    @warn_unused_result(message: "subscription must be retained to receive events")
+    public func subscribe(handler: Handler) -> Subscription {
```
Comment (LoganWright) [inline Sources/Vapor/Events/Events.swift]: I disagree Hound bot!

**Comment by LoganWright** [comment]
@tannernelson @shnhrrsn This one is ready to go.

@shnhrrsn I think this is a pretty flexible and concise design. It's pretty extensible as well, for example, we could make the syntax you suggested here:

```Swift
extension Event {
    func subscribeBool(handler: T -> Bool) {
        var subscription: Subscription?
        subscription = subscribe { data in
            let _ = subscription
            let keepOpen = handler(data)
            if !keepOpen {
                subscription = nil
            }
        }
    }
}
```

I think if users want another syntax, they can create it, but I believe structuring this way be default helps prevent memory leaks.

Maybe something like a dispose bag would be nice, for now, I'd probably use an array.

File: Sources/Vapor/Events/Events.swift [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+     - Warning: This should only be used by an event to clear the subscription on deinitialization
+     */
+    private var completion: Void -> Void = {}
+    deinit {
+        completion()
+    }
+}
+
+/**
+ This is used to contain subscriptions w/o reference
```
Comment (tannernelson) [inline Sources/Vapor/Events/Events.swift]: Can we follow the guide from http://nshipster.com/swift-documentation/ for formatting these docs?

```swift
/**
    Lorem ipsum dolor sit amet.

    @param bar Consectetur adipisicing elit.

    @return Sed do eiusmod tempor.
*/
func foo(bar: String) -> AnyObject { ... }
```

File: Sources/Vapor/Events/Events.swift [hunk 5]
```diff
@@ -0,0 +1,94 @@
+/**
+ This class is used to keep a subscription to an event active.
+
+ Set all references to `nil` to no longer receive events
+ */
+public final class Subscription {
```
Comment (tannernelson) [inline Sources/Vapor/Events/Events.swift]: Let's move these into their own files. Holder can be inside Subscription.

```swift
Subscription.Holder
```

File: Sources/Vapor/Events/Events.swift (hunk 4 above)
Comment (ketzusaka) [inline Sources/Vapor/Events/Events.swift]: @tannernelson That is the old objective-c style. Scroll down and you'll see what Logan has is the correct way to do it.

Comment (tannernelson) [inline Sources/Vapor/Events/Events.swift]: @ketzusaka Oooops. Good catch. I was talking about the indentation level and not using `*` on each line. Copied and pasted the wrong one.

File: Sources/Vapor/Events/Events.swift (hunk 5 above)
Comment (LoganWright) [inline Sources/Vapor/Events/Events.swift]: Moving to their own files unfortunately opens up some things that I'd prefer to keep private. In this case, the ability to set and use `completion`. I'd prefer to keep all of those only accessible w/in this file to help prevent bugs. Holder should be nested, I'm gonna do that now!

I can make the other stuff `internal`, but I think it should be private considering the delicate nature of the object's behavior.

Comment (tannernelson) [inline Sources/Vapor/Events/Events.swift]: `internal` should be fine. If you're working on Vapor, you'll know how this works. Or someone reviewing a PR will. Plus they might change `private` behavior soon.

Comment (LoganWright) [inline Sources/Vapor/Events/Events.swift]: 💥  Done

File: Sources/Vapor/Event/Event.swift [hunk 6]
```diff
@@ -0,0 +1,78 @@
+extension Subscription {
+    /**
+     This is used to contain subscriptions w/o reference
+     it is necessary for the reference management to work
+
+     It should only be used internally.
+     */
```
Comment (tannernelson) [inline Sources/Vapor/Event/Event.swift]: Can we tab these over like the following to match the rest of the project?

```swift
/**
    This is used to contain subscriptions w/o reference
    it is necessary for the reference management to work

    It should only be used internally.
*/
```

Comment (LoganWright) [inline Sources/Vapor/Event/Event.swift]: Surely!

File: Sources/Vapor/Event/Subscription.swift [hunk 7]
```diff
@@ -0,0 +1,16 @@
+/**
+    This class is used to keep a subscription to an event active.
+
+    Set all references to `nil` to no longer receive events
+ */
+public final class Subscription {
+    /**
+        Completion to run on deinit.
+
+        - Warning: This should only be used by an event to clear the subscription on deinitialization
```
Comment (tannernelson) [inline Sources/Vapor/Event/Subscription.swift]: hahahhahaha. you can't win

**Pull request merged**